fix: Unable to submit booking form if Attendee Phone is used and then booking Fields are saved

vijayraghav-io opened the pull request:
## What does this PR do?

- Fixes #15727 
- Fixes CAL-4025 

We now show a Badge in front of Location field to convey that it has come from the locations. The count is the number of locations used by the organizer for the event.
![image](https://github.com/user-attachments/assets/92e5bb16-84d5-4aa9-a754-37da0f6a59bd)


Also reverts #13989 but fixes the problem addressed by it in the correct way. The problems addressed:
- Ability to change the Location label.
- Ability to reorder the Location when there is just Attendee Address/Attendee Phone

Before Changes -
https://www.loom.com/share/dd1094dca81d48039de31521f4bc6561?sid=79e39a7e-fd3d-412c-a50a-ec85bf2444c4

## Mandatory Tasks (DO NOT REMOVE)

- [x] I have self-reviewed the code (A decent size PR without self-review might be rejected).
- [x] N/A - I have added a Docs issue [here](https://github.com/calcom/docs/issues/new) if this PR makes changes that would require a [documentation change](https://docs.cal.com). If N/A, write N/A here and check the checkbox.
- [x] N/A -  I confirm automated tests are in place that prove my fix is effective or that my feature works.

vijayraghav-io committed:
update for location identifier

Amit91848 reviewed (changes requested):
@vijayraghav-io this only prevents the error from being thrown. The core issue is still unresolved. It works fine until you add a label to the question, and this does nothing about it. We still need to figure out what actually is breaking it.

When you create a booking the where is no longer Attendee Address but changes to default.


https://github.com/calcom/cal.com/assets/74371312/28ed690d-9cb9-4083-81e3-9488da021795

vijayraghav-io committed:
for typecheck err

vijayraghav-io commented:
Hi @Amit91848 , it should be working fine now :
https://www.loom.com/share/cda773b1e1604cd19430581d9bed5362?sid=4d346e40-3ab4-4687-b65b-4d295b603d48

Amit91848 commented on packages/features/bookings/lib/handleNewBooking/getBookingData.ts:
```diff
@@ -68,7 +68,12 @@ export async function getBookingData<T extends z.ZodType>({
+    location:
+      typeof responses.location === "object"
+        ? responses.location?.optionValue || responses.location?.value || ""
+        : typeof responses.location === "string"
```
This is still a work around and does not address the main issue. When you change the label, it updates the type of location from radioInput to address which is what throws the error. Please fix that instead.

![Screenshot 2024-07-12 125722](https://github.com/user-attachments/assets/45805078-cdd4-417a-9bee-a5caede13148)

vijayraghav-io replied:
Yes this is expected to happen as per https://github.com/calcom/cal.com/blob/c7c5cdf133bca0393d00314227de33957fb1e594/packages/features/form-builder/FormBuilder.tsx#L245
Hence the z.union is made for location field similar to name field, Based on user input , the location can be either an object (radioinput) or a string. This was not accomodated in Booking Form

Amit91848 replied:
Ah my bad, I missed that

vijayraghav-io replied:
No issues! @Amit91848 👍

Amit91848 reviewed (dismissed):
Left a comment.

vijayraghav-io commented:
Root cause and detailed steps to recreate:

1. Create a new eventType
2. Edit the new eventType created
    1. Edit -> EventSetup (Tab) -> Location - change to ‘In Person (Attendee address)’
    <img width="958" alt="Screenshot 2024-07-12 at 12 57 31 PM" src="https://github.com/user-attachments/assets/68cba695-7acc-4b29-965a-5ee099ff3554">

    2. Things will be fine until now, and everything works just fine until you go to the ‘Advanced’ Tab and click on ‘Edit’ Button for Location inside Booking questions
    3. Once you click on Edit button. The following LOC will change type of booking field from radioinput to string
    https://github.com/calcom/cal.com/blob/c7c5cdf133bca0393d00314227de33957fb1e594/packages/features/form-builder/FormBuilder.tsx#L245
   <img width="1103" alt="Screenshot 2024-07-12 at 1 09 49 PM" src="https://github.com/user-attachments/assets/fac6ce23-bb79-471f-a70a-0b5f500dcb6f">

    4. But the booking form expects the location to still be object , Hence the input_error was occurring without much detailed message
    
3. Hence the zod-util.ts file is updated with z.union so that the location field can be either an object or a string. This accomodates the location field to be either object or string  and elimanates input_error in booking input form.
4. Also getBookingData.ts is updated to show the exact address if location is a string instead of radio input

Amit91848 commented on packages/features/bookings/lib/handleNewBooking/getBookingData.ts:
```diff
@@ -68,7 +68,12 @@ export async function getBookingData<T extends z.ZodType>({
     email: responses.email,
     guests: responses.guests ? responses.guests : [],
-    location: responses.location?.optionValue || responses.location?.value || "",
+    location:
+      typeof responses.location === "object"
+        ? responses.location?.optionValue || responses.location?.value || ""
+        : typeof responses.location === "string"
+        ? responses.location
+        : "",
```
```suggestion
    location: responses.location
      ? typeof responses.location === "object"
        ? responses.location.optionValue || responses.location.value
        : responses.location
      : "",
```

NIT: Since location can only be object or string, if not object then we shouldn't have to check if its a string.

vijayraghav-io committed:
Update packages/features/bookings/lib/handleNewBooking/getBookingData.ts

Co-authored-by: Amit Sharma <74371312+Amit91848@users.noreply.github.com>

Amit91848 commented:
Changes look good @vijayraghav-io  but would also probably need update in the docs as structure of webhooks and maybe also workflow would change for location. Since @CarinaWolli and @joeauyeung approved #13989, would like a second review from them to know if this is how we should attempt to fix this.

Normal response
`      "location": {
        "label": "location",
        "value": {
          "optionValue": "+919992929299",
          "value": "phone"
        },
        "isHidden": false
      },`
      
 Response after changing the label:
 `      "location": {
        "label": "Your Phone Number",
        "value": "+91992929292",
        "isHidden": false
      },`

vijayraghav-io committed:
Merge branch 'main' of https://github.com/calcom/cal.com into fix#15727_embedAddress

hariombalhara reviewed (dismissed):
Thanks @vijayraghav-io for the PR. But this isn't how it should be fixed, `location` prop format should stay the same.

The bug is introduced by this PR https://github.com/calcom/cal.com/pull/13989. 


The question is why it is being set as a string in the first place.

vijayraghav-io commented:
> Thanks @vijayraghav-io for the PR. But this isn't how it should be fixed, `location` prop format should stay the same.
> 
> The bug is introduced by this PR #13989.
> 
> The question is why it is being set as a string in the first place.

@hariombalhara , ok this change (changing type to string) was already there , so dint analyze why this was made.
If requirement is not to change the `location` prop format, then there has to be a different solution.
I am looking into older PRs to find why this change was made.

vijayraghav-io commented:
@hariombalhara  Analysis done:

The fix https://github.com/calcom/cal.com/pull/13989 is not good fix for https://github.com/calcom/cal.com/issues/13670
This fix tries to unhide the location question in booking questions (advanced tab) when `inPersonAttendee` location  is selected (which was originally made hidden by this PR https://github.com/calcom/cal.com/pull/7734).

Since just by unhiding - the other requirements mentioned in https://github.com/calcom/cal.com/issues/13670 , like make the field required.. are not met (in below img).
So whoever fixed might have thought that this requirements can be easily met by converting field type from `radioinput` to `string`, without may be trying (testing) to book an event with this fix. Then, the `input error` (main problem reported in this issue) would have been caught.

<img width="820" alt="Screenshot 2024-07-18 at 1 45 36 PM" src="https://github.com/user-attachments/assets/f83de890-acb3-4475-8c56-940f2ffdf3fc">

Also if user provides a custom label ("Your Address') by editing the location booking question. This was not respected (just by unhiding the bookingField).


To summarize:
1. The requirement of - hiding the location Bookingfield (or Booking question) if there is only one option (mentioned in https://github.com/calcom/cal.com/pull/7734 ) should be revisited. I guess the requirement could be like hide the location field when there is no input required (say for video apps, only link is enough without any input reqd from user).
2. If above implementation is reversed , still some updates have to be done to respect the custom label (like 'Your address') entered in Booking questions (Advanced Tab) to be shown in booking Form.

vijayraghav-io committed:
Merge branch 'main' of https://github.com/calcom/cal.com into fix#15727_embedAddress

vijayraghav-io committed:
update to use custom label for location, remove usage of hideWhenOneOption, remove string conversion of booking field

vijayraghav-io committed:
nit

vijayraghav-io committed:
nit

vijayraghav-io committed:
nit

vijayraghav-io commented:
@hariombalhara,

Updated with following changes, please check and share your comments -

1. Undone the changes from this PR - https://github.com/calcom/cal.com/pull/13989

2. Removed the usage of `hideWhenJustOneOption` property. Functionally this was used for two purposes -
    i. To remove the `location` booking field itself from Advanced Tab, when `location` had one option. 
        Actually this is not required, but hidden property on `location` booking field had to be set.
    ii. In `getBookingResponsesSchema()-> preprocess() -> schema.superRefine()`.
        This is not effective, as actually the field.hidden is set in `getAndUpdateNormalizedValues()` when there is only one non-input option.
        https://github.com/calcom/cal.com/blob/e6a16f5b5c69a4bc1a8c4b8819183b20c0ef1a21/packages/features/form-builder/FormBuilderField.tsx#L170

3. Updated to show custom label as set in `Advanced Tab` for `location booking field` in Booking Form. This will work regardless of number of location options selected.

4. Updated to set `hidden` and `required` properties of `location booking field` to be shown in `Advanced Tab` , based on the selection and number of location options in `Event Setup` tab. 
Even without this, the rendering of location fields in Booking form works fine. This is to show consistent data while editing the eventType and for the issue https://github.com/calcom/cal.com/issues/13670 , that when only `In person Attendee` location is selected, the `location booking field` should show as required in `Advanced Tab`.

vijayraghav-io commented:
@hariombalhara , this is w.r.t testing

Tested for following functionalities and is working fine :

1. Location field to be hidden in Booking Form, if there is only one non-input option. (This was already implemented)
2. The custom label entered in `Advanced Tab` (while editing the EventType) for `location Booking field` is respected and rendered in `Booking Form` while booking. This works the same irrespective of number of location options selected.
3. While editing the EventType, if `In Person Attendee` is selected as location in `Event Setup` tab, the `'location booking field` is shown in `Advanced Tab`. 
Also it is set as `required` if this is the only location option.
Will be set as `hidden`, if any conferencing app is the only option.
For all other multiple location options, default to required=false and hidden=false. And will be taken care during the booking form rendering.

hariombalhara committed:
Merge remote-tracking branch 'origin/main' into fix#15727_embedAddress

hariombalhara committed:
Use the dataStore only instead of modifying bookingFields directly from outside the FormBuilder

hariombalhara commented:
@vijayraghav-io Thanks for the patience. I have started looking into it now and changed a few things in a single commit. Let me know if it makes sense to you. 

I will be testing it more and probably add a few  tests as well before merging it.

hariombalhara commented on apps/web/components/eventtype/Locations.tsx:
```diff
@@ -268,23 +268,6 @@ const Locations: React.FC<LocationsProps> = ({
                         });
                         showToast(t("location_already_exists"), "warning");
                       }
-                      // Whenever location changes, we need to reset the locations item in booking questions list else it overflows
-                      // previously added values resulting in wrong behaviour
-                      const existingBookingFields = getValues("bookingFields");
-                      const findLocation = existingBookingFields.findIndex(
-                        (field) => field.name === "location"
-                      );
-                      if (findLocation >= 0) {
-                        existingBookingFields[findLocation] = {
-                          ...existingBookingFields[findLocation],
-                          type: "radioInput",
-                          label: "",
-                          placeholder: "",
-                        };
-                        setValue("bookingFields", existingBookingFields, {
-                          shouldDirty: true,
-                        });
-                      }
```
Reverting the code in the faulty PR. No new addition here.

FormBuilder has a dataStore property by which it is made aware of locations and then FormBuilder or the component using the FormBuilder component should do any transformations.

hariombalhara commented on packages/features/form-builder/FormBuilder.tsx:
```diff
@@ -45,6 +44,13 @@ function getCurrentFieldType(fieldForm: UseFormReturn<RhfFormField>) {
+  // 'location' must be shown as required in the UI(but at backend it is not required because not specifying it defaults to CalVideo)
+  // So, handle it in UI only instead of updating bookingFields config in getBookingFields
+  // TODO: FormBuilder must not be made aware of BookingFields, so move this "location" check to outside FormBuilder where the component is used.
+  return field.name === "location" ? true : field.required;
```
This makes a new change as well where "location" field is shown as required now. This created confusion earlier for some users.

hariombalhara commented on packages/features/form-builder/FormBuilder.tsx:
```diff
@@ -119,29 +125,37 @@ export const FormBuilder = function FormBuilder({
+              // TODO: The dataStore should itself provide the label directly
+              // This is important because FormBuilder isn't aware of what a location is. It is a generic Form Builder
+              const sourceLabel = field.getOptionsAt === "locations" ? "Location" : field.getOptionsAt;
+              options.forEach((option) => {
+                sources.push({
+                  id: option.value,
+                  label: sourceLabel,
+                  type: "system",
+                });
+              });
```
![image](https://github.com/user-attachments/assets/4e36a0dc-960e-4fbb-a3fa-ff1d1d942c0c)
This shows a badge that this field is coming due to locations. This is similar to what we do on fields added through workflows

hariombalhara commented on packages/features/form-builder/FormBuilder.tsx:
```diff
@@ -119,29 +125,37 @@ export const FormBuilder = function FormBuilder({
+            const doesFirstOptionHaveInput = !!firstOptionInput;
+            // If there is only one option and it doesn't have an input required, we don't show the Field for it.
+            // Because booker doesn't see this in UI, there is no point showing it in FormBuilder to configure it.
+            if (field.hideWhenJustOneOption && numOptions <= 1 && !doesFirstOptionHaveInput) {
```
We need to still use `hideWhenJustOneOption` because it is used in getBookingResponsesSchema as well.

Also, this makes the code handle any other radioInput kind of custom field(when we allow that in future)

hariombalhara commented on packages/features/form-builder/FormBuilderField.tsx:
```diff
@@ -54,6 +54,7 @@ export const FormBuilderField = ({
   const { t } = useLocale();
   const { control, formState } = useFormContext();
 
+  const labelBeforeNormalize = field.label;
```
Why are we using the `labelBeforeNormalize`? What is the difference here?

vijayraghav-io replied:
This is required to respect the custom label entered for location in BookingQuestions, because the `getAndUpdateNormalizedValues()` fn will modify the field.label to default label `location`.

hariombalhara replied:
I have fixed the logic so that we use field.label only. The issue was because we were setting field.label to empty string for radioInput type field(e.g. location) when it has just one option and that is an input.

Now, we are not modifying that, and introduced a separate property noLabel and doing some prop drilling.

vijayraghav-io commented:
> @vijayraghav-io Thanks for the patience. I have started looking into it now and changed a few things in a single commit. Let me know if it makes sense to you.
> 
> I will be testing it more and probably add a few tests as well before merging it.

Thankyou! @hariombalhara , i am looking through the changes and replied to comment.

hariombalhara committed:
Dont unset the actual label, instead have a different property noLabel for this

hariombalhara commented on packages/features/form-builder/Components.tsx:
```diff
@@ -398,17 +406,17 @@ export const Components: Record<FieldType, Component> = {
 
-      const getCleanLabel = (option: { label: string; value: string }): string | JSX.Element => {
-        if (!option.label) {
+      const getCleanLabel = (label: string): string | JSX.Element => {
```
We need just the `label`. This way it works even if I have just the label.

hariombalhara commented on packages/features/form-builder/Components.tsx:
```diff
@@ -398,17 +408,24 @@ export const Components: Record<FieldType, Component> = {
 
-      const getCleanLabel = (option: { label: string; value: string }): string | JSX.Element => {
-        if (!option.label) {
+      const didUserProvideLabel = (
```
This is who we decide if the organizer has changed the label and if the label has been changed by the organizer, we show the organizer's label instead of the first option label.

hariombalhara commented on packages/features/form-builder/FormBuilder.tsx:
```diff
@@ -244,16 +257,7 @@ export const FormBuilder = function FormBuilder({
                       data-testid="edit-field-action"
                       color="secondary"
                       onClick={() => {
-                        const fieldToEdit = field;
-                        // For radioInput type, when there's only one option, the type and required takes the only first options values
-                        if (fieldToEdit.type === "radioInput" && options.length === 1) {
-                          fieldToEdit.type =
-                            fieldToEdit.optionsInputs?.[options[0].value].type || fieldToEdit.type;
-                          fieldToEdit.required =
-                            fieldToEdit.optionsInputs?.[options[0].value].required || fieldToEdit.required;
-                        }
-
-                        editField(index, fieldToEdit);
```
Reverted code from faulty PR.

hariombalhara commented on packages/features/form-builder/FormBuilder.tsx:
```diff
@@ -119,29 +125,37 @@ export const FormBuilder = function FormBuilder({
+            if (field.hideWhenJustOneOption && numOptions <= 1 && !doesFirstOptionHaveInput) {
               return null;
             }
-            let fieldType = fieldTypesConfigMap[field.type];
-            let isRequired = field.required;
-            // For radioInput type, when there's only one option, the type and required takes the first options values
-            if (field.type === "radioInput" && options.length === 1) {
-              fieldType = fieldTypesConfigMap[field.optionsInputs?.[options[0].value].type || field.type];
-              isRequired = field.optionsInputs?.[options[0].value].required || field.required;
-            }
```
Reverted code from faulty PR.

hariombalhara commented on packages/features/form-builder/FormBuilderField.tsx:
```diff
@@ -227,7 +244,7 @@ export const ComponentForField = ({
   if (componentConfig.propsType === "text") {
     return (
-      <WithLabel field={field} readOnly={readOnly}>
+      <WithLabel field={field} readOnly={readOnly} noLabel={noLabel}>
```
Pass on `noLabel` directly instead of setting field.label to empty string because that would mean that someone wanting to use the field.label still won't have it.

hariombalhara committed:
Merge branch 'main' into fix#15727_embedAddress

CarinaWolli reviewed (commented):
Tested it and everything worked great 👍
